WI #616: Addition of five columns related to `Pricing Currency` (#848)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/columns/pricingcurrencyeffectivecost.md

# Pricing Currency Effective Cost

The Pricing Currency Effective Cost represents the cost of the [*charge*](#glossary:charge) after applying all reduced rates, discounts, and the applicable portion of relevant, prepaid purchases (one-time or recurring) that covered this charge, as denominated in [Pricing Currency](#pricingcurrency).  Pricing Currency Effective Cost can be denominated in [Billing Currency](#billingcurrency) by multiplying by the [Pricing to Billing Exchange Rate](#pricingtobillingexchangerate).  This allows the practitioner to perform a conversion from either 1) a [*national currency*](#glossary:nationalcurrency) to a [*virtual currency*](#glossary:virtualcurrency) (e.g. tokens to USD), or 2) one national currency to another (e.g. EUR to USD).

The PricingCurrencyEffectiveCost column adheres to the following requirements:

> * PricingCurrencyEffectiveCost MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports pricing and billing in different currencies.

Author: Agreement per discussion in Mar 24 Maintainers.

```suggestion
* PricingCurrencyEffectiveCost presence in a [*FOCUS dataset*](#glossary:FOCUS-dataset) is defined as follows:
  * PricingCurrencyEffectiveCost MUST be present in a FOCUS dataset when the provider supports prices in virtual currency and publishes unit prices exclusive of discounts.
  * PricingCurrencyEffectiveCost is RECOMMENDED to be present in a FOCUS dataset when the provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
  * PricingCurrencyEffectiveCost MAY be present in a FOCUS dataset in all other cases.
```

# Pricing Currency Effective Cost

> The Pricing Currency Effective Cost represents the cost of the [*charge*](#glossary:charge) after applying all reduced rates, discounts, and the applicable portion of relevant, prepaid purchases (one-time or recurring) that covered this charge, as denominated in [Pricing Currency](#pricingcurrency).  Pricing Currency Effective Cost can be denominated in [Billing Currency](#billingcurrency) by multiplying by the [Pricing to Billing Effective Exchange Rate](#pricingtobillingeffectiveexchangerate).  This allows the practitioner to perform a conversion from either 1) a [*national currency*](#glossary:nationalcurrency) to a [*virtual currency*](#glossary:virtualcurrency) (e.g. tokens to USD), or 2) one national currency to another (e.g. EUR to USD).

Reviewer: **Disclaimer:** I am not 100% sure what PricingCurrencyEffectiveCost represents, so I might be asking a silly question :)

**Question:** Does the following hold true?

`PricingToBillingEffectiveExchangeRate `= `EffectiveCost` / `PricingCurrencyEffectiveCost
`

If so, should we rephrase the following sentence by mentioning Effective Cost? 

> Pricing Currency Effective Cost can be denominated in [Billing Currency](#billingcurrency) by multiplying by the [Pricing to Billing Effective Exchange Rate](#pricingtobillingeffectiveexchangerate).
>
> **Expressed as a formula**

> * PricingCurrencyEffectiveCost denominated in BillingCurrency = PricingCurrencyEffectiveCost * PricingToBillingEffectiveExchangeRate

PricingCurrencyEffectiveCost denominated in BillingCurrency, mentioned in the above sentence  represents Effective Cost, right?

And perhaps we should also add the following requirements:

> * The product of PricingCurrencyEffectiveCost and [PricingToBillingEffectiveExchangeRate](#pricingtobillingeffectiveexchangerate) MUST match the [EffectiveCost](#effectivecost) when ChargeClass is not "Correction".
> * Discrepancies in PricingCurrencyEffectiveCost and EffectiveCost MAY exist when ChargeClass is "Correction".

Author: Discussed in Mar 28 TF3.  Will add these normative requirements to this column, and will add verbiage to the overview section to describe that math to get to Effective Cost.

Author: Discussed further with @ijurica.  Will leave out the second suggested normative requirement because they will often be discrepant as a matter of course, being denominated in different currencies.

Author: Met with @ijurica and added verbiage to address this.  Resolving.

The PricingCurrencyEffectiveCost column adheres to the following requirements:

* PricingCurrencyEffectiveCost presence in a [*FOCUS dataset*](#glossary:FOCUS-dataset) is defined as follows:
  * PricingCurrencyEffectiveCost MUST be present in a FOCUS dataset when the provider supports virtual currencies (e.g. credits, tokens).

>   * PricingCurrencyEffectiveCost MUST be present in a FOCUS dataset when the provider presents prices and usage in different national currencies (e.g. prices in USD and usage in EUR).

Author: Per suggestion from @UdamDewaraja [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848#discussion_r2004634306), which he suggested before I renamed the column (my apologies for not committing this before I did that).  Also, TIL that [e.g. takes a comma but i.e. does not](https://english.stackexchange.com/questions/16172/should-i-always-use-a-comma-after-e-g-or-i-e)!

```suggestion
  * PricingCurrencyEffectiveCost MUST be present in a FOCUS dataset when the provider uses different national currencies for pricing and billing (e.g., prices in USD and bills in EUR).
```

  * PricingCurrencyEffectiveCost MAY be present in a FOCUS dataset in all other cases.
* PricingCurrencyEffectiveCost MUST be of type Decimal.
* PricingCurrencyEffectiveCost MUST conform to [NumericFormat](#numericformat) requirements.
* PricingCurrencyEffectiveCost MUST NOT be null.
* PricingCurrencyEffectiveCost MUST be a valid decimal value.

> * PricingCurrencyEffectiveCost MUST be denominated in the [PricingCurrency](#pricingcurrency).

Reviewer 2: ```suggestion
* PricingCurrencyEffectiveCost MUST be denominated in the [PricingCurrency](#pricingcurrency).
* PricingCurrencyEffectiveCost MUST be 0 in the event of prepaid purchases or purchases that are applicable to previous usage.

## Column ID

PricingCurrencyEffectiveCost

## Display Name

Pricing Currency Effective Cost

## Description

> The cost of the [*charge*](#glossary:charge) as denominated in [Pricing Currency](#pricingcurrency).

Author: As suggested by @UdamDewaraja [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848/files/e45f1d1f32affa42ce89a0ab3c676aeeec505c94#r2004653227) before I performed the column rename per a separate suggestion.
```suggestion
The cost of the [*charge*](#glossary:charge) after applying all reduced rates, discounts, and the applicable portion of relevant, prepaid purchases (one-time or recurring) that covered this charge, as denominated in [Pricing Currency](#pricingcurrency).
```

The PricingCurrencyEffectiveCost column adheres to the following requirements:

* PricingCurrencyEffectiveCost presence in a [*FOCUS dataset*](#glossary:FOCUS-dataset) is defined as follows:
  * PricingCurrencyEffectiveCost MUST be present in a FOCUS dataset when the provider supports prices in virtual currency and publishes unit prices exclusive of discounts.
  * PricingCurrencyEffectiveCost is RECOMMENDED to be present in a FOCUS dataset when the provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
  * PricingCurrencyEffectiveCost MAY be present in a FOCUS dataset in all other cases.
* PricingCurrencyEffectiveCost MUST be of type Decimal.
* PricingCurrencyEffectiveCost MUST conform to [NumericFormat](#numericformat) requirements.
* PricingCurrencyEffectiveCost MUST NOT be null.
* PricingCurrencyEffectiveCost MUST be a valid decimal value.
* PricingCurrencyEffectiveCost MUST be denominated in the [PricingCurrency](#pricingcurrency).

> * PricingCurrencyEffectiveCost MUST be 0 in the event of prepaid purchases or purchases that are applicable to previous usage. 

Author: Applying suggestion from @ijurica per [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848/files#r2018676712):

```suggestion
* PricingCurrencyEffectiveCost MUST be 0 in the event of prepaid purchases or purchases that are applicable to previous usage. 
* The product of PricingCurrencyEffectiveCost and [PricingToBillingEffectiveExchangeRate](#pricingtobillingeffectiveexchangerate) MUST match the [EffectiveCost](#effectivecost) when ChargeClass is not "Correction".
```

## Column ID

PricingCurrencyEffectiveCost

## Display Name

Pricing Currency Effective Cost

## Description

> The cost of the [*charge*](#glossary:charge) after applying all reduced rates, discounts, and the applicable portion of relevant, prepaid purchases (one-time or recurring) that covered this charge, as denominated in [Pricing Currency](#pricingcurrency).

Reviewer 3: ```suggestion
The cost of the *charge* after applying all reduced rates, discounts, and the applicable portion of relevant, prepaid purchases (one-time or recurring) that covered this charge, as denominated in *Pricing Currency*.
```

> 
> ## Usability Constraints
> 
> **Aggregation:** Given that Pricing Currency Effective Cost can contain a combination of national and/or virtual currencies, it is necessary to group by the [Pricing Currency](#pricingcurrency) column before performing any Pricing Currency Effective Cost aggregations.  Other cost columns such as [Effective Cost](#effectivecost) state all values in a unified [Billing Currency](#billingcurrency) and therefore do not require such a grouping.

Reviewer 3: ```suggestion
```
This either would need to be said in any cost column OR we leave it implicit. The last statement is incorrect. EffectiveCost (or any other cost column that's provided in BillingCurrency) has the same limitation. across different providers both sets of 'Cost' columns can have >1 currency and therefore would need to be grouped by the respective currency.

## Content Constraints

|    Constraint   |      Value              |
|:----------------|:------------------------|
| Column type     | Metric                  |

> | Feature level   | Conditional             |

Reviewer 3: ```suggestion
| Feature level   | Recommended             |
```

Looks like InvoiceId has the wrong value for the Feature level as well.

Author: The definition of Conditional:

> * If the existence of a column is described as MUST with conditions of when it applies, then the feature level is designated as 'Conditional'.

That's the case here.  This column has a presence of MUST in certain conditions.  Can you tell me more about why you believe this should be Recommended?

Reviewer: Regarding the InvoiceId column, the Feature level constraint was addressed (set to Recommended) in the PR that was merged into the working draft last week.

Reviewer 3: Yes, I missed that. This is a must where as the ones I was pattern matching to were not stating MUST with a condition. And yes, I was looking at an old version of InvoiceId. Thanks -- resolving.
